feat: Transactional Produces to Command Topic

stevenpyzhang opened the pull request:
### Description 
Protocol
```
If a statement needs to be distributed (in DistributingExecutor):
Initialize a new producer with a fixed value (based on ksql.service.id) for transactions.id. All KSQL request handlers must always use the same value for transactions.id. This ensures that Request Handlers that already started a transaction are unable to commit once a given producer's transaction is in-flight.
Start a transaction 
Wait for the command topic consumer to consume up to the last offset (we need to ask kafka for the last offset after starting the transaction.)
Validate the request
Produce the command to the command topic
Commit the transaction
```
We want to ensure that every command written to the commandTopic is being validated only after all previous commands have been processed in order to have the most up to date snapshot of the KsqlEngine for validation purposes. 

Some notes
- I removed the existing CommandProducer from CommandTopic.java
- An entire transaction is in DistributingExecutor
- A new consumer is created every time we `waitForCommandConsumer` in order to get the end offset of the command topic, this can't be done with the existing Consumer in CommandTopic.java because it's running in the CommandRunner thread, which is separate from the thread handling requests in KsqlResource (consumers are not thread-safe)
-new configs added to improve durability of command topic, set directly in code unless otherwise specified.
```
//for durability of command topic
ksql.internal.topic.replication.factor=3 //passed in through configs
ksql.internal.topic.min.insync.replicas=2 //passed in through configs
unclean.leader.election.enable=false

producer.acks=all // only for the command topic producer

//to enable transactions
isolation.level="read_commited" //consumer
transactional.id=[ksql_server_id] //producer
```

These changes also require additional ACLs to be set including 
```
WRITE TRANSACTIONAL_ID "<ksql.service.id>": A transactional producer which has its transactional.id set requires this privilege.

DESCRIBE TRANSACTIONAL_ID "<ksql.service.id>": This applies only on transactional producers and checked when a producer tries to find the transaction coordinator.

DESCRIBE TOPIC "__consumer_offsets" "__transaction_state"
```

Follow up to this PR:
https://github.com/confluentinc/ksql/issues/3795
https://github.com/confluentinc/ksql/issues/3768

Transactional Producer API Docs
https://kafka.apache.org/10/javadoc/index.html?org/apache/kafka/clients/producer/KafkaProducer.html

Klip introducing transactions
https://cwiki.apache.org/confluence/display/KAFKA/KIP-98+-+Exactly+Once+Delivery+and+Transactional+Messaging#KIP-98-ExactlyOnceDeliveryandTransactionalMessaging-TestPlan

More on Transactional Messaging in Kafka
https://cwiki.apache.org/confluence/display/KAFKA/Transactional+Messaging+in+Kafka

### Testing done 
Existing tests fixed, may need a new integration test to ensure functionality.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

stevenpyzhang commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/RequestHandler.java:
```diff
@@ -102,14 +114,19 @@ public KsqlEntityList execute(
         customExecutors.getOrDefault(statementClass, distributor);
 
+    if (executor instanceof DistributingExecutor) {
+      ((DistributingExecutor) executor).setTransactionManager(producerTransactionManager);
```
I wasn't sure of the best way to handle passing the producerTransactionManager to DistributingExecutor. The interface for StatementExecutor.execute() is shared among many of the non-distributed executors so I didn't think it was appropriate to modify it to pass along the manager.

purplefox replied:
I think customExecutors never contains a DistributingExecutor, so if you don't find the executor in customExecutors you could just create a new DistributingExecutor and you can pass the producerTransactionManager into the constructor of that, so you wouldn't need a setTransactionManager method.

stevenpyzhang replied:
There already exists a DistributingExecutor which is set as the executor for the current statement if the statementClass isn't found in the customExecutor.

https://github.com/confluentinc/ksql/blob/master/ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/RequestHandler.java#L47

rodesai replied:
+1 to @purplefox's suggestion. We just need to rejigger the code a bit. `DistributingExecutor` doesn't need to implement `StatementExecutor`, and it's `execute` method can take a `TransactionalProducer` as a parameter. Then, `RequestHandler` can just instantiate a wrapper object that _does_ implement `StatementExecutor`, or just use a lambda:

```
final StatementExecutor<T> executor = customExecutors.getOrDefault(
    statementClass,
    (stmt, props, ctx, svcCtx -> distributor.execute(stmt, props, ctx, svcCtx, transactionalProducer));
```

stevenpyzhang replied:
^ that works, I'll use that then.

stevenpyzhang commented on ksql-test-util/src/main/java/io/confluent/ksql/test/util/EmbeddedSingleNodeKafkaCluster.java:
```diff
@@ -470,6 +470,10 @@ private Properties buildBrokerConfig(final String logDir) {
     config.put(KafkaConfig.LogRetentionTimeMillisProp(), -1);
     // Stop logs marked for deletion from being deleted
     config.put(KafkaConfig.LogDeleteDelayMsProp(), Long.MAX_VALUE);
+    // Set to 1 because only 1 broker
```
Without setting these, this error appears for some integration tests
```
ERROR [KafkaApi-0] Number of alive brokers '1' does not meet the required replication factor '3' for the transactions state topic (configured via 'transaction.state.log.replication.factor'). This error can be ignored if the cluster is starting up and not all brokers are up yet. (kafka.server.KafkaApis:74)```

stevenpyzhang commented:
> Could you elaborate a bit on the above statement? I'm trying to understand why this is important. Perhaps you could illustrate with an example?

Currently, before we produce commands to the command topic, we do validation on the statement to ensure it can be executed properly. And when we're actually executing the statement on a KSQL server after consuming it from the command topic, we're also doing these validations. 

There's two main forms of validations being done, internal state and external state. For example, a `Create Stream FooStream...` statement, Internal state validation would be ensuring no stream already exists with the name `FooStream` while external state validation would be checking to see if the underlying Kafka Topic `Foo` exists.

We're trying to move away from doing validation when executing the statement. 
The work being done with Klip-6 https://github.com/confluentinc/ksql/blob/master/design-proposals/klip-6-execution-plans.md eliminates the need for external validations. This PR would eliminated the need for doing internal state validations when executing.

A simple example of something that this code would affect is currently if two KSQL servers concurrently try to produce the same CREATE STREAM FOO to the command topic, both would be enqueued. One of the statements would fail when it's actually executed. Although in this example the final state of the KSQL server is fine (a Stream named FOO is created), It would be better if we can ensure every statement written to the command topic at offset n has been validated against a KSQL internal state that has processed all statements up to n-1 (the internal KSQL state is most up to date).

stevenpyzhang commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/ProducerTransactionManager.java:
```diff
@@ -0,0 +1,161 @@
+    commandProducer.beginTransaction();
+  }
+
+  public void waitForCommandRunner() {
```
Wasn't exactly sure of the best way to do this so suggestions would be appreciated.

purplefox commented:
Thanks Steven, I think I have a clearer picture of the motivations here now.

> It would be better if we can ensure every statement written to the command topic at offset n has been validated against a KSQL internal state that has processed all statements up to n-1 (the internal KSQL state is most up to date).

This seems to imply that every time we add anything to the command topic we must first wait for all outstanding commands to be processed locally.

I think this would mean we're going to incur, _at least_ a round trip time to the Kafka broker for each command, i.e. our throughput we will be limited by the latency of ksql<->kafka. If that's measured in milliseconds it means our theoretical maximum throughput will be in the 100s of transactions of second, which seems very low. 

I'm wondering if we really want to impose such a bottleneck on the system?

purplefox commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/KsqlResource.java:
```diff
@@ -229,6 +252,7 @@ public Response handleKsqlStatements(
       return ErrorResponseUtil.generateResponse(
           e, Errors.badRequest(e));
     } catch (final Exception e) {
+      e.printStackTrace();
```
Best to remove this

stevenpyzhang replied:
woops, yea i'll get rid of it, left over from my debugging

vcrfxia commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/ProducerTransactionManager.java:
```diff
@@ -0,0 +1,161 @@
+
+    try {
+      int retries = 0;
+      while (commandRunner.getNumCommandProcessed() < endOffset) {
```
Is this assuming that every offset in the command topic corresponds to a command? I thought that with transactions these need not necessarily be in sync since transactions write additional messages into the topic (which take up offsets but aren't commands), though I'm having a surprisingly tough time finding docs about this. Would be good to have someone else confirm or deny.

stevenpyzhang replied:
When I was testing this locally, I was able to make several transactions in a row without this method getting stuck. If additional messages were taking up offsets, I'd hit the retry limit and have an exception thrown which didn't happen. So that behavior makes me think it doesn't effect commandConsumer.endOffsets() but someone more familiar with this should chime in. I can't seem to find any documentation about these additional messages either.

apurvam replied:
Every transaction will write at least one control message to each partition in the transaction. This control message (commit or abort message) will have an offset. I'm actually surprised that your local testing worked. 

Anothre risk is if we have aborted transactions: and there is a chance of these if an active produce gets fenced by a newer instance coming on line. In this case, the number of valid (non-aborted commands) will always be divergent from the end offsets. 

The docs on how this works can be found here https://cwiki.apache.org/confluence/display/KAFKA/KIP-98+-+Exactly+Once+Delivery+and+Transactional+Messaging

stevenpyzhang replied:
I ended up reusing the functionality we already have in place for commandStore.ensureConsumedPast() which uses offsets. position() will get the next offset to fetch so this will ignore the control messages

rodesai commented:
> I'm wondering if we really want to impose such a bottleneck on the system?

The reason I'm not concerned about this is that we only need to follow the protocol implemented here for statements that need to be logged to the command topic (CREATE STREAM, CREATE TABLE, CREATE STREAM AS, CREATE TABLE AS, DROP, TERMINATE). We don't expect a significant rate of these on any KSQL cluster.

vinothchandar commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestApplication.java:
```diff
@@ -646,7 +668,8 @@ private static void maybeCreateProcessingLogStream(
 
-    commandQueue.enqueueCommand(configured.get());
+    commandQueue.enqueueCommand(configured.get(), transactionalProducer);
+    transactionalProducer.commit();
```
is the exception handling and the abortTransaction() handled inside `enqueueCommand` > `Transactional Producer::send`? could we avoid calling commit() for an aborted transaction ?

rodesai commented:
> > I'm wondering if we really want to impose such a bottleneck on the system?
> 
> The reason I'm not concerned about this is that we only need to follow the protocol implemented here for statements that need to be logged to the command topic (CREATE STREAM, CREATE TABLE, CREATE STREAM AS, CREATE TABLE AS, DROP, TERMINATE). We don't expect a significant rate of these on any KSQL cluster.

Ah I see the current implementation does this wait for all commands. We should fix that.

stevenpyzhang commented:
> > > I'm wondering if we really want to impose such a bottleneck on the system?
> > 
> > 
> > The reason I'm not concerned about this is that we only need to follow the protocol implemented here for statements that need to be logged to the command topic (CREATE STREAM, CREATE TABLE, CREATE STREAM AS, CREATE TABLE AS, DROP, TERMINATE). We don't expect a significant rate of these on any KSQL cluster.
> 
> Ah I see the current implementation does this wait for all commands. We should fix that.

Should the TransactionalProducer be moved to DistributingExecutor then? It would help get rid of that injection pattern I currently have and it makes sense since DistributingExecutor would be the only one that uses it. 

I think the main problem though is that the transaction needs to be started before the RequestValidator validation happens. If there's CREATE STREAM, CREATE TABLE, etc... 
mixed in with non-distributed statements like SHOW TOPICS, I think it only makes sense to wait beforehand for the commandRunner and then validate all the statements.

rodesai commented:
> Should the TransactionalProducer be moved to DistributingExecutor then? It would help get rid of that injection pattern I currently have and it makes sense since DistributingExecutor would be the only one that uses it.
> 
> I think the main problem though is that the transaction needs to be started before the RequestValidator validation happens. If there's CREATE STREAM, CREATE TABLE, etc...
> mixed in with non-distributed statements like SHOW TOPICS, I think it only makes sense to wait beforehand for the commandRunner and then validate all the statements.

I think the current structure is reasonable (other than the way the producer is injected - but I'll leave that for the code comments) - the distributing executor doesn't need to know about validation. It would definitely be worthwhile to optimize away creating the producer/consumer, and checking offsets for statements that don't need to distribute to the command topic though.

stevenpyzhang commented:
One option for getting rid of injecting the TransactionalProducer into DistributingExecutor could be to have it implement a DistributedStatementExecutor interface which extends StatementExectuor. DistributedStatementExecutor interface would have an execute function with the required arg types for passing in the producer.

stevenpyzhang commented:
>optimize away creating the producer/consumer
@rodesai I thought during a discussion we had about implementation details, you mentioned that we needed to create a new Producer for each transaction since there could be multiple threads on a server trying to produce to the command topic. I believe the consumers for getting offset could be optimized to just one consumer that's shared among all the TransactionalProducers (passed to each new TransactionalProducer from the factory class).

rodesai commented:
> > optimize away creating the producer/consumer
> 
> @rodesai I thought during a discussion we had about the implementation details of this, you mentioned that we needed to create a new Producer for each transaction since there could be multiple threads on a server trying to produce to the command topic. I believe the consumers for getting offset could be optimized to just one consumer that's shared among all the TransactionalProducers (passed to each new TransactionalProducer from the factory class).

I just mean optimizing it away when handling requests that don't need a transaction (e.g. `list streams;`, or a pull query). For requests that include statements that need to be distributed let's just optimize for keeping the code as simple as possible, and not worry about creating a new consumer/producer each time for these.

rodesai commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/TransactionalProducer.java:
```diff
@@ -0,0 +1,161 @@
+
+    try {
+      int retries = 0;
+      while (commandRunner.getNumCommandProcessed() < endOffset) {
```
I don't think we want to use the number of commands processed here. We should just be able to ask the command runner which offset its consumed up to. After all, this is really what we care about and protects us from bugs with bookkeeping the number of commands that we've run.

stevenpyzhang replied:
uses offsets now

rodesai commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/DistributingExecutor.java:
```diff
@@ -64,18 +67,26 @@ public DistributingExecutor(
+      final QueuedCommandStatus queuedCommandStatus =
+          commandQueue.enqueueCommand(injected, transactionalProducer);
+
       final CommandStatus commandStatus = queuedCommandStatus
```
We could also re-run validation for logged statements as part of execute. So `DistributingExecutor` would just run the whole protocol (init txn, validate, log cmd, commit). It's a bit wasteful to re-run validation, but I think this is the simplest way to get this PR unblocked.

stevenpyzhang replied:
> - run the validate-execute-commit loop for each statement one-by-one (rather than validating them all up front).

I think we have to go with this option. The first option could potentially return misleading results in the event of an abortTransaction();

```
//starting from no streams
CREATE STREAM foo1(age BIGINT) WITH (KAFKA_TOPIC='foo',  VALUE_FORMAT='JSON');
CREATE STREAM foo2(age BIGINT) WITH (KAFKA_TOPIC='foo',  VALUE_FORMAT='JSON');
show streams;
```
If validations passed but when trying to commit the 2 records, it fails and the transaction is aborted, those streams wouldn't be created, but the show streams in that batch of statements would show that Stream foo1 and foo2 were created since it would be successfully executed against a sandboxEngine and then the `show streams` is run against a sandboxEngine with those two streams. However, running show streams again would show no streams have been created.

stevenpyzhang replied:
> - run the validate-execute-commit loop for each statement one-by-one (rather than validating them all up front).

I think there's also issues with this. With how the code is currently designed, a KsqlEntityList is only returned if all the statements are successfully executed.

```
ksql> CREATE TABLE foo400(age BIGINT) WITH (KAFKA_TOPIC='foo',  VALUE_FORMAT='JSON');
>CREATE TABLE foo200(age BIGINT) WITH (KAFKA_TOPIC='foo',  VALUE_FORMAT='JSON');
>Show Tables;

CREATE TABLE foo400(age BIGINT) WITH (KAFKA_TOPIC='foo',  VALUE_FORMAT='JSON');
 Message       
---------------
 Table created 
---------------

CREATE TABLE foo200(age BIGINT) WITH (KAFKA_TOPIC='foo',  VALUE_FORMAT='JSON');
 Message       
---------------
 Table created 
---------------

Show Tables;
 Table Name | Kafka Topic | Format | Windowed 
----------------------------------------------
 FOO200     | foo         | JSON   | false    
 FOO400     | foo         | JSON   | false    
----------------------------------------------
ksql> 
ksql> CREATE TABLE foo100(age BIGINT) WITH (KAFKA_TOPIC='foo',  VALUE_FORMAT='JSON');
>CREATE TABLE foo200(age BIGINT) WITH (KAFKA_TOPIC='foo',  VALUE_FORMAT='JSON');
>Show Tables;
Cannot add table 'FOO200': A table with the same name already exists
ksql> show tables;

 Table Name | Kafka Topic | Format | Windowed 
----------------------------------------------
 FOO100     | foo         | JSON   | false    
 FOO200     | foo         | JSON   | false    
 FOO400     | foo         | JSON   | false    
----------------------------------------------
ksql> 
```

The code would need to be refactored to be able to handle processing individual commands (which should probably be done before implementing the TransactionalProduces)

rodesai replied:
> This option could potentially return misleading results in the event of an abortTransaction();
In this case we'd just fail the whole request - so no response to `list streams;`

rodesai replied:
> The code would need to be refactored to be able to handle processing individual commands (which should probably be done before implementing the TransactionalProduces)

I'm not sure I follow. It's possible today for any command to fail when executing, in which case we just fail the whole command. We could do the same thing if validation inside the transaction fails. I agree it's better to return the partial response with the results for what we could successfully execute. My point is that this is orthogonal to what we're trying to do here.

stevenpyzhang replied:
> I agree it's better to return the partial response with the results for what we could successfully execute. My point is that this is orthogonal to what we're trying to do here.

I also think it would be good to have a partial response, but it's not a deal breaker for me. If the team is OK with the behavior I outlined here https://github.com/confluentinc/ksql/pull/3660#discussion_r341442966 then I'm on board.

stevenpyzhang replied:
I ended up going with  

Running the protocol on each statement one-by-one (re running validation on distributed statements)

rodesai replied:
Yeah doing the validation in `DistributedExecutor` was just meant to be an immediate-term thing so we can get this change merged without blocking on deciding how we really want the validation to be done. I'm personally OK with validating-executing each statement one by one. This is the way it was done at first. At some point @big-andy-coates changed it to validate everything up-front. It would be good to get his perspective on why this was done.

stevenpyzhang commented:
**Updates to PR:**

New Configs
```
//for durability of command topic
ksql.streams.replication.factor=3
ksql.streams.producer.acks=all
ksql.streams.topic.min.insync.replicas=2
unclean.leader.election.enable=false

//to enable transactions
isolation.level="read_commited" //consumer
transactional.id=[ksql_server_id]
```

* Added new configs ^
* Transactional Producer waiting for CommandRunner to consume up to latest offset - uses similar structure to functionality for commandSequenceNumber
* Don't create a TransactionalProducer if the incoming request only contains non-distributed statements (pull queries, list topics, ...etc). This will get addressed in a follow up PR.
* ExceptionHandling for the transaction - an entire transaction is handled inside DistributingExecutor.execute() which is wrapped in a try catch
* Changed protocol from a batched transaction to single statement transactions

The new protocol is:
```
	Create and initialize TransactionalProducer
	Validate all the statements at once
	for each statement:
		if non distributed, execute
		if distributed
			Start transaction
			Wait for CommandRunner consumer to process all records present before start of transaction
			Validate the statement
			Send (returns QueuedCommandStatus) and Commit the record
			Try and Wait forQueueCommandStatus
		
		If exception while executing a statement, return error message specific to that statement. 
		
```

One UX point is that some early statements may be committed, but a later one fails to be committed (since statements aren't batch committed to the command topic). We'd return an error message for that failed statement, but not any partial success responses 

```
ksql> CREATE TABLE foo100(age BIGINT) WITH (KAFKA_TOPIC='foo',  VALUE_FORMAT='JSON');
>CREATE TABLE foo200(age BIGINT) WITH (KAFKA_TOPIC='foo',  VALUE_FORMAT='JSON');
>Show Tables;
>CREATE TABLE foo300(age BIGINT) WITH (KAFKA_TOPIC='foo',  VALUE_FORMAT='JSON');
>CREATE TABLE foo400(age BIGINT) WITH (KAFKA_TOPIC='foo',  VALUE_FORMAT='JSON'); // fails to be committed
>CREATE TABLE foo500(age BIGINT) WITH (KAFKA_TOPIC='foo',  VALUE_FORMAT='JSON');
>CREATE TABLE foo600(age BIGINT) WITH (KAFKA_TOPIC='foo',  VALUE_FORMAT='JSON');
>Show Tables;
Could not write the statement 'CREATE TABLE foo400(age BIGINT) WITH (KAFKA_TOPIC='foo',  VALUE_FORMAT='JSON');' into the command topic: Some exception
Caused by: Some exception
ksql> show tables;

 Table Name | Kafka Topic | Format | Windowed 
----------------------------------------------
 FOO100     | foo         | JSON   | false    
 FOO200     | foo         | JSON   | false    
 FOO300     | foo         | JSON   | false    
----------------------------------------------
```

stevenpyzhang commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/TransactionalProducerFactory.java:
```diff
@@ -0,0 +1,57 @@
+import java.util.Objects;
+import org.apache.kafka.clients.producer.ProducerConfig;
+
+public class TransactionalProducerFactory {
```
commandStore would need to keep track of the commandRunner in order to pass it to each TransactionalProducer in `createTransactionalProducer`. I'm not sure how exactly that would work out because the commandStore is created first in RestApp and then the commandRunner.

rodesai replied:
Why can't we reuse the existing mechanism in `CommandStore` to do this? (see the implementation of `completeSatisfiedSequenceNumberFutures`). Then the producer factory has no dependency on `CommandRunner`.

stevenpyzhang replied:
The CommandStore version introduces a race condition. The futures for the sequenceNumbers are completed before the commands are actually executed in commandRunner. There's a bit of lag between consuming and actually executing the command and updating the metastore. Also, the CommandRunner could get stuck executing consumed commands, but ensureConsumedPast() in CommandStore wouldn't block the TransactionalProducer.waitForConsumer() call and then the transaction would proceed on an out of sync metastore.

stevenpyzhang replied:
After thinking about it and playing around with the existing functionality of `CommandStore.ensureComsumedPast()`, it fits with what I was trying to achieve by duplicating the code. I removed the code from CommandRunner and the factory class. I'm not sold on dropping `enqueueCommand()` from commandQueue though since the `CommandStore.commandStatusMap` is being updated in `enqueueCommand()`.

rodesai replied:
> I'm not sold on dropping enqueueCommand() from commandQueue though since the CommandStore.commandStatusMap is being updated in enqueueCommand()

Is the concern that you won't be able to get to the map from the queue? In this way of doing things the implementation of the producer interface could be _inside_ CommandStore, so that shouldn't be a problem. So something like:

```
interface TransactionalProducer {
    send()
}

interface CommandQueue {
    getProducer()
}

class CommandStore {
    Producer getProducer() {
        return new TopicProducer();
    }
    class TopicProducer implements TransactionalProducer {
        send() { ... }
    }
}
```

This way, the underlying queue is still encapsulated away.

stevenpyzhang replied:
Mind if I leave this for a future PR? I think this PR has gone through enough refactors already.

stevenpyzhang commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestApplication.java:
```diff
@@ -646,7 +668,8 @@ private static void maybeCreateProcessingLogStream(
 
-    commandQueue.enqueueCommand(configured.get());
+    commandQueue.enqueueCommand(configured.get(), producerTransactionManager);
+    producerTransactionManager.commit();
```
Transactions have been isolated to DistributingExecutor and will be aborted if an exception is thrown.

stevenpyzhang commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/ProducerTransactionManager.java:
```diff
@@ -0,0 +1,161 @@
+      }
+      throw new RuntimeException(e.getCause());
+    } catch (final InterruptedException e) {
+      throw new RuntimeException(e);
```
the exception should bubble up and get handled in DistributingExecutor where the transaction would be aborted

agavra commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/TransactionalProducerImpl.java:
```diff
@@ -0,0 +1,170 @@
+    waitForConsumer();
+  }
+
+  public void waitForConsumer() {
```
can this be made private? having it public makes it a little dangerous because if you call it outside of a transaction it doesn't really guarantee anything

stevenpyzhang replied:
done

agavra commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/TransactionalProducerFactory.java:
```diff
@@ -0,0 +1,69 @@
+    );
+  }
+
+  public TransactionalProducer createProducerTransactionManager() {
```
nit:
```suggestion
  public TransactionalProducer createTransactionalProducer() {
```

stevenpyzhang commented on ksql-common/src/main/java/io/confluent/ksql/util/KsqlConfig.java:
```diff
@@ -506,6 +509,12 @@ private static ConfigDef buildConfigDef(final ConfigGeneration generation) {
+            KSQL_INTERNAL_TOPIC_MIN_INSYNC_REPLICAS_PROPERTY,
+            Type.SHORT,
+            (short) 1,
+            ConfigDef.Importance.LOW,
```
done

stevenpyzhang commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/TransactionalProducerImpl.java:
```diff
@@ -0,0 +1,170 @@
+    try {
+      final long endOffset = getEndOffset();
+      
+      // timeout hardcoded for now
```
nope it was a leftover

stevenpyzhang commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/TransactionalProducerImpl.java:
```diff
@@ -0,0 +1,170 @@
+  }
+
+  private void initialize() {
+    commandConsumer.assign(Collections.singleton(commandTopicPartition));
```
made `initialize` public

stevenpyzhang commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/TransactionalProducerImpl.java:
```diff
@@ -0,0 +1,170 @@
+    }
+  }
+
+  public void abort() {
```
I removed abort from the send.

stevenpyzhang commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/DistributingExecutor.java:
```diff
@@ -83,6 +110,8 @@ public DistributingExecutor(
       ));
     } catch (final Exception e) {
+      transactionalProducer.abort();
+      transactionalProducer.close();
```
Added finally blocks. Just to help keep track of everything, there are 3 places where transactions are being used.

Terminate path in KsqlResource
Ksql path in KsqlResource

Creating the processing log in KsqlRestApplication

I removed the close call here as the close should occur in KsqlResource in that finally block

stevenpyzhang commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/CommandRunner.java:
```diff
@@ -204,6 +211,34 @@ private void terminateCluster(final Command command) {
+          "Error waiting for commandRunner to process commandTopic up to offset "
+              + seqNum, e.getCause());
+    } catch (final TimeoutException e) {
+      throw new TimeoutException(
```
Added the original error message to the string for the new TimeoutException. Also updated the commandStore version of this.

stevenpyzhang commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestApplication.java:
```diff
@@ -624,29 +644,34 @@ private static void maybeCreateProcessingLogStream(
-    final Supplier<ConfiguredStatement<?>> configured = () -> ConfiguredStatement.of(
-        statement, Collections.emptyMap(), ksqlConfig);
-
+    final TransactionalProducer transactionalProducer =
```
https://github.com/confluentinc/ksql/issues/3768

stevenpyzhang commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/TransactionalProducerImpl.java:
```diff
@@ -0,0 +1,170 @@
+    } catch (final InterruptedException e) {
+      throw new RuntimeException(e);
+    } catch (KafkaException e) {
+      commandProducer.abortTransaction();
```
`begin()` is called in DistributingExecutor.execute()
The entire protocol for begin transaction -> validate statement -> send to command topic -> commit transaction happens in that function. Since it's all wrapped in a try catch, if any exception is thrown in DistributingExecutor.execute(), any existing transaction will be aborted.

rodesai commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/DistributingExecutor.java:
```diff
@@ -38,41 +43,63 @@
+      transactionalProducer.begin();
+      
+      // Don't perform validation on Terminate Cluster statements
+      if (statement.getStatement() instanceof TerminateCluster) {
```
shouldn't this be `if (!(statement.getStatement() instanceof TerminateCluster))`

stevenpyzhang replied:
Good catch, although this actually doesn't work. The TerminateCluster statement is actually transformed into a TerminateQuery{queryId=CLUSTER} statement after being prepared by the engine. I'm going to need to rethink how to do this.

rodesai commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/TransactionalProducer.java:
```diff
@@ -0,0 +1,34 @@
+import io.confluent.ksql.rest.server.computation.Command;
+import org.apache.kafka.clients.producer.RecordMetadata;
+
+public interface TransactionalProducer extends Cloneable {
```
Should be `Closeable`

rodesai commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/CommandRunner.java:
```diff
@@ -204,6 +211,38 @@ private void terminateCluster(final Command command) {
     log.info("The KSQL server was terminated.");
   }
 
+  public void ensureProcessedPastOffset(final long seqNum, final Duration timeout)
```
Why did we have to add this? Why not just use `CommandStore.ensureConsumedPast`

stevenpyzhang commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/CommandRunner.java:
```diff
@@ -204,6 +211,10 @@ private void terminateCluster(final Command command) {
     log.info("The KSQL server was terminated.");
   }
 
+  public long getNumCommandProcessed() {
```
Ended up using `CommandStore.ensureConsumedPast()`

stevenpyzhang commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/TransactionalProducerImpl.java:
```diff
@@ -0,0 +1,170 @@
+  }
+
+  public void close() {
+    commandConsumer.close();
```
Added guava `Closer` to this method. Also added some logging in case the TransactionalProducer fails to close.

agavra commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/TransactionalProducer.java:
```diff
@@ -0,0 +1,59 @@
+import org.eclipse.jetty.util.IO;
+
+
+public interface TransactionalProducer extends Closeable {
```
I think this could benefit from a lot more documentation (and merge it with the documentation that's in the implementation class). Namely there's a lot of questions that go un-answered that would be really helpful (at least for me) to review and understand the code.

Overall:
- What is the expected usage pattern for this class?
- Can it be shared across threads?

About `begin`:
- Is this method blocking?
- If a producer on server A calls `begin()` before server B calls `begin()`, what happens to B? Is B still allowed to proceed?

About `send`:
- When is the send resolved?
  - What happens if I call a send without ever calling an abort/commit (perhaps I crash in between)?
  - What happens if I call multiple sends in one transaction?
- What happens if a send fails (i.e. throws an exception)?
- What happens if I call a send before I call begin or after I call commit/abort/close?

About commit/abort:
- What happens if I call commit/abort without calling send or before calling start()?
- When I call abort, what happens to the events I called send for?
- Does calling abort/commit end my transaction? If so what is `close` used for?

Basically I encourage you to think about all of these scenarios (and come up with your own as well) and document them thoroughly. It gets especially complicated when you have multiple servers trying to do these at the same time.

stevenpyzhang replied:
I could add it to the code inline, but I'll probably be copying a lot over from the Kafka Transactional Producer documentation. https://kafka.apache.org/10/javadoc/index.html?org/apache/kafka/clients/producer/KafkaProducer.html

Maybe it would be just easier to link to that documentation? Most of the cases you listed out our documented there.

stevenpyzhang replied:
I was also thinking about just switching over to using the KafkaProducer interface for the implementation.
https://cwiki.apache.org/confluence/display/KAFKA/KIP-98+-+Exactly+Once+Delivery+and+Transactional+Messaging#KIP-98-ExactlyOnceDeliveryandTransactionalMessaging-ProducerAPIchanges

There are some additional functions in this interface I don't think we currently need though. The only real nuance between the default KafkaProducer interface and the implementation is that we add an additional block in beginTransaction() for the commandRunner to catch up and process all the records. If we still want to have our own interface for this producer, would copying over the Kafka documentation be sufficient?

agavra replied:
Getting these APIs right is pretty hard - if you think we don't necessarily need to have our own interface, i'd much rather just use the vanilla kafka one. The "waiting" behavior is a little unclear to me (e.g. my questions about `begin` weren't easily answered looking at the KafkaProducer code).

rodesai replied:
We should not do that, as it leaks the fact that Kafka is the underlying queue.

agavra replied:
If we're adding transactionality, especially tied to the semantics that Kafka exposes, I don't think this is a bad thing.

stevenpyzhang replied:
@agavra I added a lot more documentation now to the interface. Lmk if there's anything that I missed. A lot of the edge cases brought up go back to doing the transaction process out of order in some fashion. All the methods will throw a KafkaException if called out of order.

agavra replied:
Thanks @stevenpyzhang! I'll take a look at the new documentation and add specific comments if I have questions

rodesai replied:
My opinion is that it's worthwhile to keep the rest endpoints decoupled from how we distribute/handle commands. It's already useful for testing, and gives us more flexibility to change things if we want to down the line. I'm not going to die on this hill though. But if we're going that route, I'd like to see two things:

- Just use Kafka's producer interface. There's no value adding and maintaining our own interface here, but plenty of cost (keeping semantics consistent, keeping documentation up to date)
- Let's push the usage of the interface as close to the usage of CommandStore as possible. So create the producer right before beginning each transaction and don't pass it around everywhere (e.g. just create it from `DistributingExecutor.execute`).

stevenpyzhang replied:
> Let's push the usage of the interface as close to the usage of CommandStore as possible. So create the producer right before beginning each transaction and don't pass it around everywhere (e.g. just create it from DistributingExecutor.execute).

I guess this also solves the problem of not creating the producer for a set of non-distributed statement. Isn't the overhead going to be a lot for creating a new TransactionalProducer(KafkaProducer and KafkaConsumer) for every single distributed statement?

agavra replied:
I'm not so worried about overheads for distributed commands - it's probably not perceivable to the human at the end of the CLI, and nobody is submitting these commands programmatically by the thousands

stevenpyzhang replied:
If a user were to only ever submit single statements at a time, the overhead is around the same, this would only really be noticeable for large batches of distributed statements.

stevenpyzhang replied:
> nobody is submitting these commands programmatically by the thousands

Who knows, given enough users, maybe one of them will come up with some bizarre use case for that. But yea, this makes sense currently. I can refactor to keep everything in the DistributingExecutor.

agavra replied:
Isn't there a default limit of like 10 running queries at one time anyway? 😂

rodesai replied:
I'd mostly be worried about the added latency. But that would only be for batch statements, which are probably not the norm anyway. I'd favor code quality here unless we discover it's really a problem.

stevenpyzhang replied:
Once I make this refactor I can try creating 100 streams in one request

stevenpyzhang replied:
100 create streams takes around 10 seconds which is very reasonable.

agavra commented on config/ksql-production-server.properties:
```diff
@@ -74,9 +74,11 @@ ksql.streams.producer.delivery.timeout.ms=2147483647
 
-# For better fault tolerance and durability, set the replication factor for the KSQL
-# Server's internal topics. Note: the value 3 requires at least 3 brokers in your Kafka cluster.
+# For better fault tolerance and durability, set the replication factor and minimum insync replicas
```
If you're up for it, I think we should split all of the new configuration work out into a separate PR and get that one merged quickly - you'll be able to make some progress on that front and it'll also make this one easier to review.

stevenpyzhang replied:
I think it's not worth the effort to break it out. The configuration for the command topic effects 4 files, which I don't think adds too much complexity to a review.

```
ksql-production-server.properties
KsqlConfig
KsqlInternalTopicUtil and test file
```

agavra replied:
just a thought for untying these - but if there's no rush then I don't mind :)

stevenpyzhang commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestApplication.java:
```diff
@@ -623,27 +630,46 @@ private static void maybeCreateProcessingLogStream(
+        .collect(Collectors.toSet());
+
     if (!config.getBoolean(ProcessingLogConfig.STREAM_AUTO_CREATE)
-        || !commandQueue.isEmpty()) {
```
This check doesn't work anymore because there could be an aborted processing log statement in the command topic.

stevenpyzhang replied:
reverted.

rodesai commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestApplication.java:
```diff
@@ -623,27 +630,46 @@ private static void maybeCreateProcessingLogStream(
     try {
+      transactionalProducer.initTransactions();
+      transactionalProducer.beginTransaction();
+  
```
We might wind up writing multiple CS messages for the log stream now, which is never supposed to happen. When we fix up the command runner to no longer skip commands, this will result in it getting stuck on the repeats. I think it would be better for this PR to keep the old behavior of checking the end offset (after starting the txn), and accept that if we have to abort we won't create the log stream. When you refactor, this should get fixed anyway.

stevenpyzhang replied:
I'll revert and track this fix here. https://github.com/confluentinc/ksql/issues/3768

rodesai commented on ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/computation/DistributingExecutorTest.java:
```diff
@@ -30,6 +30,7 @@
 import io.confluent.ksql.KsqlExecutionContext;
 import io.confluent.ksql.exception.KsqlTopicAuthorizationException;
 import io.confluent.ksql.metastore.MetaStore;
+import io.confluent.ksql.parser.KsqlParser.ParsedStatement;
```
Can you add a specific test case to this file that ensures that the write protocol is followed? It should ensure that first we start the transaction, then verify, then write out the statement, and then commit.

stevenpyzhang replied:
added one

stevenpyzhang commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/DistributingExecutor.java:
```diff
@@ -16,63 +16,95 @@
       final ConfiguredStatement<Statement> statement,
+      final ParsedStatement parsedStatement,
       final Map<String, Object> mutableScopedProperties,
+      final String sql,
```
I'll use `parsedStatement.getStatementText()`

stevenpyzhang commented on ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestApplication.java:
```diff
@@ -628,22 +637,39 @@ private static void maybeCreateProcessingLogStream(
       return;
```
Only commit() and abort() raise the offset I believe, I moved only the empty check after the transaction starts.
